Stage the March 2023 Starship blogs (#6067)
sourcegraph/about: content/blogposts/2023/sourcegraph-5.0.md

---
title: "Sourcegraph 5.0"
publishDate: 2023-03-23T10:00-07:00
description: "Announcing Sourcegraph 5.0. The latest release includes 20+ updates and introduces Cody, the AI programming assistant."
tags: [blog, release]
slug: "release/5.0"
published: true
heroImage: https://storage.googleapis.com/sourcegraph-assets/blog/release-post/5.0/sourcegraph-5-0-hero.png
socialImage: https://storage.googleapis.com/sourcegraph-assets/blog/release-post/5.0/sourcegraph-5-0-hero.png
changelogItems:
---

> We all know we're headed towards a future where there are a lot more people coding and even more code generated, much of it written by AI. With today's tools, this will be a mess. But with the right tools, this can lead to the most sustained period of human advancement.

Reviewer: ```suggestion
We're headed towards a future with a lot more people and AIs writing a lot more code. With today's tools, this will be a mess. But with the right tools, this can lead to the most sustained period of human advancement.
```

* **The graph of code**, which means all the relevant code plus code navigation, dep graph, ownership, runtime, logging, tracing, deployment, and more.
* **Workflows** to help devs write a new feature faster, fix bugs/vulns faster, and do refactors/migrations faster.
* **Large language models** that read the relevant parts of the code graph to combine general intuition with specific data to power these workflows.

"Source_graph_"? "The graph of code"? Yes, we've known the graph of code would be incredibly valuable ever since we started Sourcegraph in 2013. We didn't predict something like LLMs would come so quickly, but we've been hoping for this moment for the last 10 years.

Today, we're making another big step toward this vision with Sourcegraph 5.0, which introduces:
* **Cody**, our AI assistant that feeds context from Sourcegraph's code graph to the LLM to answer your questions about your own entire codebase;
* **Own**, which adds code ownership information to the code graph;
* **App**, which runs Sourcegraph as a local desktop app so any dev can start using it quickly;

> * And much more. 

Reviewer: ```suggestion
* and much more. 
```

since it is a continuation of the sentence spread across the bullets

Read on to discover everything included in Sourcegraph 5.0.
<br/>

<Badge link="/cody" text="AI" color="blue" size="small" />
#### Cody, your code-aware programmer's assistant

> Cody is an AI coding assistant that can find, explain, and write code. Cody can answer questions about both general programming topics and your specific codebase from right inside your editor. Cody is powered by Sourcegraph Code Intelligence, so unlike other coding assistant, it knows about your codebase and not just about open source code.

Reviewer: ```suggestion
Cody is an AI coding assistant that can find, explain, and write code. Cody can answer questions about both general programming topics and your specific codebase from right inside your editor. Cody is powered by Sourcegraph's graph of code intelligence, so unlike other coding assistants, it knows about your codebase and not just about open source code.
```

> We're releasing Cody as experimental in 5.0, with general availability coming in a future release. [Learn more about Cody](https://about.sourcegraph.com/cody) and [sign up for access to Cody.](https://sourcegraph.typeform.com/cody-signup)

Reviewer: ```suggestion
We're releasing Cody as experimental in 5.0, with general availability coming in a future release. [Learn more about Cody](https://about.sourcegraph.com/cody) and [sign up for access to Cody](https://sourcegraph.typeform.com/cody-signup).
```

<br/>

<Badge link="/own" text="Code Ownership" color="green" size="small" />
#### Sourcegraph Own preview

Sourcegraph Own manages code ownership for your entire codebase across every repository and code host. React to incidents faster and increase productivity by making it easy to find who owns any code quickly.

> We are launching a preview of Sourcegraph Own as an experimental feature and will iterate over the coming months to deliver our full roadmap vision to every customer. [Read more in the full announcement for Sourcegraph Own](https://about.sourcegraph.com/blog/inroducing-sourcegraph-own).

Reviewer: ```suggestion
We are launching a preview of Sourcegraph Own as an experimental feature and will iterate over the coming months to deliver our full roadmap vision to every customer. [Read more in the full announcement for Sourcegraph Own.](https://about.sourcegraph.com/blog/inroducing-sourcegraph-own)
```

<Video 
  source={{
    webm: 'blog/announcing-own',
    mp4: 'blog/announcing-own'
  }}
  loop={true}
  title="An example of looking at ownership data in Sourcegraph"
/>

<a href="https://docs.sourcegraph.com/own" className="not-italic flex items-center mb-sm">Docs<OpenInNewIcon className="ml-xxs" size={18} /></a>
<br/>

<Badge link="/code-search" text="Code Search" color="cerise" size="small" />
#### Improved code exploration experience

In 5.0, we’re introducing a number of exciting code exploration experience improvements.

> First and foremost, we've shipped the CodeMirror editor as the default blob viewer. This new editor is much more efficient with large files, making it easier and faster than ever to explore code. It also has great extensibility potential, enabling us to ship more IDE-like code exploration features in the future.

Reviewer: ```suggestion
We shipped the CodeMirror editor as the default blob viewer. This new editor is much more efficient with large files, making it easier and faster than ever to explore code. It also has great extensibility potential, enabling us to ship more IDE-like code exploration features in the future.
```

We've also enhanced the blob view by adding keyboard navigation. This means you can move around your codebase quickly and efficiently without ever taking your hands off the keyboard. We’ve also added keyboard shortcuts for switching between the code editor, files, and symbols trees. You can turn this on using the `blob-page-switch-areas-shortcuts` feature flag.

We're also shipping file and symbol trees with drastically improved accessibility and a slick browsing experience.

Plus, we've made a number of improvements to the repo and directory pages. We're now rendering the README file in the directory so you can get a quick overview of what's inside. We've also added a linear file browser, which shows diff stats over the past month, so you can quickly see what's changed recently. And we're now showing recent commits and contributors, so you can get a sense of who's been working on the codebase.

> Finally, we've made some improvements to the git blame decorations. We've improved the design and now show the file in the previous revision. Where possible, we're linking to GitHub issues and pull requests, so you can quickly jump to relevant discussions.

Reviewer: ```suggestion
Finally, we've made some improvements to the Git blame decorations. We've improved the design and now show the file in the previous revision. Where possible, we're linking to GitHub issues and pull requests, so you can quickly jump to relevant discussions.
```

All of these improvements add up to a more streamlined, efficient, and powerful code exploration experience. We're excited to see how they'll help our users better understand and work with the codebase!

<br/>
<Video
  source={{
    webm: 'blog/keyboard-nav',
    mp4: 'blog/keyboard-nav'
  }}
  loop={true}
  title="New keyboard navigation experience in Sourcegraph"
/>
<br/><br/>

<Badge link="/code-search" text="Code Search" color="cerise" size="small" />
#### A completely re-imagined search input

> The search input has been redesigned to greatly improve usability. New contextual suggestions help users learn the Sourcegraph query language as they search. These powerful suggestions react to the user’s current query to suggest related filters. How suggestions are displayed has been unified across contexts and filters, and the history mode has been integrated directly into input. Lastly, improved and expanded keyboard shortcuts also make navigation much easier.

Reviewer: ```suggestion
The search input has been redesigned to greatly improve usability. New contextual suggestions help users learn the Sourcegraph query language as they search. These powerful suggestions react to the user’s current query to suggest related filters. How suggestions are displayed has been unified across contexts and filters, and the history mode has been integrated directly into input. Finally, improved and expanded keyboard shortcuts also make navigation much easier.
```

<br/>
<div className="mb-sm">
  <Figure
    src="https://storage.googleapis.com/sourcegraph-assets/blog/5.0/reimagined-search-input.png"
    alt="The new search UI"
    catpion="A re-imagined search input."
  />
</div><br/>

<Badge link="/code-search" text="Code Search" color="cerise" size="small" />
#### Intelligent search ranking

> Previously, Sourcegraph ordered search results in a simple way that could make it hard to find the most relevant files. In version 4.6, we're releasing two significant improvements to search ranking:

Reviewer: ```suggestion
Previously, Sourcegraph ordered search results in a simple way that could make it hard to find the most relevant files. In version 5.0, we're releasing two significant improvements to search ranking:
```

We are launching a preview of Sourcegraph Own as an experimental feature and will iterate over the coming months to deliver our full roadmap vision to every customer. [Read more in the full announcement for Sourcegraph Own](https://about.sourcegraph.com/blog/inroducing-sourcegraph-own).

<Video 
  source={{
    webm: 'blog/announcing-own',
    mp4: 'blog/announcing-own'
  }}
  loop={true}
  title="An example of looking at ownership data in Sourcegraph"
/>

<a href="https://docs.sourcegraph.com/own" className="not-italic flex items-center mb-sm">Docs<OpenInNewIcon className="ml-xxs" size={18} /></a>
<br/>

<Badge link="/code-search" text="Code Search" color="cerise" size="small" />
#### Improved code exploration experience

In 5.0, we’re introducing a number of exciting code exploration experience improvements.

First and foremost, we've shipped the CodeMirror editor as the default blob viewer. This new editor is much more efficient with large files, making it easier and faster than ever to explore code. It also has great extensibility potential, enabling us to ship more IDE-like code exploration features in future.

> We've also enhanced the blob view by adding keyboard navigation. This means you can move around yiour codebase quickly and efficiently without ever having to take your hands off the keyboard. We’ve also added keyboard shortcuts for switching between the code editor, files, and symbols trees. You can turn this on using the `blob-page-switch-areas-shortcuts` feature flag.

Reviewer 2: ```suggestion
We've also enhanced the blob view by adding keyboard navigation. This means you can move around your codebase quickly and efficiently without ever having to take your hands off the keyboard. We’ve also added keyboard shortcuts for switching between the code editor, files, and symbols trees. You can turn this on using the `blob-page-switch-areas-shortcuts` feature flag.
```

<br />

<Badge text="Code Insights" link="/code-insights" color="green" size="small" />
#### Improved Code Insights support for instances with a large number of repositories

Code insights provide precise answers about the trends and composition of your codebase, tranforming it into a queryable database.  Previously, users had to create insights over either a few named repositories or all of their repositories, and running insights over all repositories could often take an impractical amount of time to process. Now, with the new repository selection, users can target their insights to the exact scope of repositories that is relevant to them. Additionally, administrators can now monitor the processing of insights and, when necessary, reprioritize or retry them from within the site admin section.

<a href="https://docs.sourcegraph.com/code_insights" className="not-italic flex items-center mb-sm">Docs<OpenInNewIcon className="ml-xxs" size={18} /></a>
<br />

<Badge link="https://docs.sourcegraph.com/admin/external_service" text="Code Hosts" color="violet" size="small" />
#### Impoved Gerrit support with user permissions

> Gerrit connections now have their own dedicated code host connection option as opposed to the “Generic Git Host” connection that had to be used before. Along with this, Gerrit connections can now also enable authorization, which will mark all repositories from that connection as private and require users to authenticate using Gerrit credentials and verify that they have permission to view that repository on Gerrit if they want to be able to view it on Sourcegraph.

Reviewer 3: I'm not sure about

> which will mark all repositories from that connection as private

@pjlast @kopancek -- is that the correct way to phrase it?

Reviewer 3: Yeah, that sounds good to me!

Author: ```suggestion
Gerrit connections now have their own dedicated code host connection option as opposed to the “Generic Git Host” connection that had to be used before. Gerrit connections can also be configured to require user authentication, in which case users will only be able to view repositories by providing their own Gerrit credentials.
```

> Gerrit connections now have their own dedicated code host connection option as opposed to the “Generic Git Host” connection that had to be used before. Along with this, Gerrit connections can now also enable authorization, which will mark all repositories from that connection as private and require users to authenticate using Gerrit credentials and verify that they have permission to view that repository on Gerrit if they want to be able to view it on Sourcegraph.

Reviewer 4: ```suggestion
Gerrit connections now have their own dedicated code host connection option as opposed to the “Generic Git Host” connection that had to be used before. Along with this, Gerrit connections can also be configured to require user authentication, which will require users to provide Gerrit credentials in order to search and browse Gerrit code.
```

#### Improved support for the Azure DevOps code host

Sourcegraph customers will now be able to bring their Azure DevOps repositories into their Sourcegraph instance. Sourcegraph will sync specified repositories from the customers’ Azure DevOps organizations or projects also respect and enforce permissions from Azure DevOps, ensuring that only users with access rights to repositories are allowed to access them. Batch Changes will now also support Azure DevOps code hosts and repositories.

<a href="https://docs.sourcegraph.com/admin/external_service/azuredevops" className="not-italic flex items-center mb-sm">Docs<OpenInNewIcon className="ml-xxs" size={18} /></a>
<br />

<Badge link="https://docs.sourcegraph.com/admin/external_service" text="Code Hosts" color="violet" size="small" />
#### Improved rate limiting for GitHub and GitLab

We've made changes to how we handle GitHub and GitLab rate limits. Previously, all GitHub requests would be limited to 5,000 requests per hour (i.e. even if there were 1,000 users on the instance, the total number of requests would not exceed 5,000/hour).

> Now we use feedback from the code host to do rate limiting, and each individual user has their own rate limit. This greatly speeds up our permissions syncing process. Customers might see an increase in the number of requests to the code host, but these will not exceed the code host's own imposed rate limits.

Reviewer 3: It's the same thing :)

```suggestion
Now we use feedback from the code host to do rate limiting using each individual user's own rate limit. This greatly speeds up our permissions syncing process. Customers might see an increase in the number of requests to the code host, but these will not exceed the code host's own imposed rate limits.
```

<br/>

<Badge link="/batch-changes" text="Batch Changes" color="blue" size="small" />
#### Integrate Batch Changes with other tools with outgoing webhooks

> Batch Changes allows to apply and track code changes across many repositories and codehosts. In many pipelines, opening and merging changesets requires interacting with other tools: ticketing, custom automation for reviews, etc. Now outgoing webhooks for Batch Changes let you send changeset events out to third party systems.

Reviewer 5: ```suggestion
Batch Changes enables you to apply and track code changes across many repositories and code hosts. In many pipelines, opening and merging changesets requires interacting with other tools: ticketing, custom automation for reviews, etc. Now, outgoing webhooks for Batch Changes let you send batch change and changeset events out to third party systems automatically.
